text/stateful-resources-one-pager.md

```diff
@@ -0,0 +1,258 @@
+
+```json
+{
+  "AWS::S3::Bucket": ["BucketName"],
+  "AWS::DynamoDB::Table": ["TableName"],
+}
+```
+
+Using this file, we can generate the correct constructor, for example:
+
+```ts
+export class CfnBucket extends cdk.CfnResource implements cdk.IInspectable {
+
+  constructor(scope: cdk.Construct, id: string, props: CfnBucketProps = {}) {
+      super(scope, id, {
+        stateful: true,
+        replacedIf: ["BucketName"],
+        type: CfnBucket.CFN_RESOURCE_TYPE_NAME,
+        properties: props
+      });
+  }
+}
+```
+
+### Synthesis
+
+Now that all L1 resources are correctly marked and can be inspected, we add
+logic to the synthesis process to write this metadata to the cloud assembly.
+For example, a stack called `my-stack`, defining an `S3` bucket with
+id `MyBucket`, will contain the following in `manifest.json`:
+
+```json
+{
+  "metadata": {
+    "/my-stack/MyBucket/Resource": [
+      {
+        "stateful": true,
+        "replacedIf": ["BucketName"],
+      }
+    ],
+  }
+}
+```
+
+### Deployment
+
+During deployment, we inspect the diff and identify which changes are about
+to occur on stateful resources in the stack. This closes the loop and should
+allow for the [customer experience](#desired-customer-experience) we described.
+
+## Q & A
+
+### Why do we need the API change given we have the curated file?
+
+Its true that the existence of the file allows us to codegen marked L1 resources
+without needing to make changes to the input API. However, this won't
+cover the use-case of **stateful custom resources**. As we know, custom resources may
+also be stateful. This scenario is in fact what sparked this discussion.
+
+> See [s3: toggling off auto_delete_objects for Bucket empties the bucket](https://github.com/aws/aws-cdk/issues/16603)
+
+Since every [CustomResource] eventually instantiates a `CfnResource`,
+changing this API will also make sure custom resources are marked appropriately.
+
+In addition, enforcing this on the API level will also be beneficial for third-party
+constructs that might be extending from `CfnResource`.
+
+[CustomResource]: https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/core/lib/custom-resource.ts#L123
+
+### What about CDK Pipelines?
+
+This document only refers to a deployment workflow using `cdk deploy`. I have yet
+to research how this would be done in CDK pipelines. My initial thoughts are to somehow incorporate
+`cdk diff` in the pipeline, along with a manual approval stage.
+
+### Would this have prevented [#16603](https://github.com/aws/aws-cdk/issues/16603)?
+
+Yes. If the API proposed here existed, we would have passed `stateful: true` in the
+implementation of the `autoDeleteObjects` property. Then, when the resource was
+removed due to the toggle, we would identify this as a removal of a stateful resource, and block it.
+
+### Curating a list of stateful resources is a maintenance burden - can we do it differently?
```

Comment:
Modifying the file should be straightforward and infrequent. So I think it's only a burden to the extent that people can forget to update it when creating a new stateful resource. Is there any mechanism to remind developers to check this (like a lint rule)?